beluga_amcl/src/amcl_nodelet.cpp

```diff
@@ -245,6 +245,7 @@ auto AmclNodelet::make_particle_filter(const nav_msgs::OccupancyGrid::ConstPtr&
 }
 
 void AmclNodelet::map_timer_callback(const ros::TimerEvent&) {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Uh, was this the culprit? I thought we were not using any `GetMap` services.

Reply:
We don't use it, but we instantiate the service in this test:
https://github.com/Ekumen-OS/beluga/blob/7df85b4ad96f5550bd1b138bea2cd8e5f9884d20/beluga_amcl/test/test_amcl_nodelet.cpp#L348-L363

And the backtrace shows we reach that part of the code.

```bash
Thread 6 "test_amcl_nodel" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f02366e2700 (LWP 1928)]
0x000055b62596123e in std::vector<Sophus::SE2<double, 0>, std::allocator<Sophus::SE2<double, 0> > >::size (this=0x18) at /usr/include/c++/9/bits/stl_vector.h:916
916           { return size_type(this->_M_impl._M_finish - this->_M_impl._M_start); }
(gdb) bt
#0  0x000055b62596123e in std::vector<Sophus::SE2<double, 0>, std::allocator<Sophus::SE2<double, 0> > >::size (this=0x18) at /usr/include/c++/9/bits/stl_vector.h:916

#1  0x000055b62595ad80 in beluga::TupleContainer<beluga::Vector, std::tuple<Sophus::SE2<double, 0>, beluga::Numeric<double, beluga::WeightTag, void> > >::size (this=0x0)
    at /ws/install/beluga/include/beluga/beluga/containers/tuple_vector.hpp:94
#2  0x00007f023c653090 in beluga_amcl::AmclNodelet::initialize_from_estimate (this=0x55b6265f5cf0, estimate={...}) at /ws/src/beluga/beluga_amcl/src/amcl_nodelet.cpp:560
#3  0x00007f023c6504c0 in beluga_amcl::AmclNodelet::handle_map_with_default_initial_pose (this=0x55b6265f5cf0, map=...)
    at /ws/src/beluga/beluga_amcl/src/amcl_nodelet.cpp:359
#4  0x00007f023c64ec20 in beluga_amcl::AmclNodelet::map_timer_callback (this=0x55b6265f5cf0) at /ws/src/beluga/beluga_amcl/src/amcl_nodelet.cpp:260
#5  0x00007f023c747f60 in boost::_mfi::mf1<void, beluga_amcl::AmclNodelet, ros::TimerEvent const&>::operator() (this=0x55b6265f8148, p=0x55b6265f5cf0, a1=...)
    at /usr/include/boost/bind/mem_fn_template.hpp:165
```

I guess there are some destruction shenanigans happening here.

Reply:
Wait, we do have a map server, right? 
https://github.com/Ekumen-OS/beluga/blob/7df85b4ad96f5550bd1b138bea2cd8e5f9884d20/beluga_amcl/test/test_amcl_nodelet.cpp#L105C5-L105C16

Reply:
So, basically, I think the race was between this write, read and completing the test right after the read:

- Write: https://github.com/Ekumen-OS/beluga/blob/7df85b4ad96f5550bd1b138bea2cd8e5f9884d20/beluga_amcl/src/amcl_nodelet.cpp#L294
- Read: https://github.com/Ekumen-OS/beluga/blob/7df85b4ad96f5550bd1b138bea2cd8e5f9884d20/beluga_amcl/test/test_amcl_nodelet.cpp#L66
- Test: https://github.com/Ekumen-OS/beluga/blob/7df85b4ad96f5550bd1b138bea2cd8e5f9884d20/beluga_amcl/test/test_amcl_nodelet.cpp#L362

Adding the lock let's the initialization complete before finishing the test.